ethcore/src/state.rs

```diff
@@ -39,41 +39,97 @@ pub struct ApplyOutcome {
 pub type ApplyResult = Result<ApplyOutcome, Error>;
 
 #[derive(Debug)]
-enum AccountEntry {
+enum AccountData {
 	/// Contains account data.
 	Cached(Account),
 	/// Account has been deleted.
-	Killed,
 	/// Account does not exist.
 	Missing,
 }
 
-impl AccountEntry {
-	fn is_dirty(&self) -> bool {
+impl AccountData {
+	/// Clone dirty data into a new `AccountData`.
+	fn clone_dirty(&self) -> AccountData {
 		match *self {
-			AccountEntry::Cached(ref a) => a.is_dirty(),
-			AccountEntry::Killed => true,
-			AccountEntry::Missing => false,
+			AccountData::Cached(ref acc) => AccountData::Cached(acc.clone_dirty()),
+			AccountData::Missing => AccountData::Missing,
 		}
 	}
 
+	/// Clone account entry data that needs to be saved in the snapshot.
+	/// This includes basic account information and all locally modified storage keys
+	fn clone_for_snapshot(&self) -> AccountData {
+		match *self {
+			AccountData::Cached(ref acc) => AccountData::Cached(acc.clone_dirty()),
+			AccountData::Missing => AccountData::Missing,
+		}
+	}
+}
+
+#[derive(Eq, PartialEq, Clone, Copy, Debug)]
+enum AccountState {
+	Clean,
+	Dirty,
+	Commited,
+}
+
+#[derive(Debug)]
+struct AccountEntry {
+	account: AccountData,
+	state: AccountState,
+}
+
+impl AccountEntry {
+	fn is_dirty(&self) -> bool {
+		self.state == AccountState::Dirty
+	}
+
 	/// Clone dirty data into new `AccountEntry`.
 	/// Returns None if clean.
 	fn clone_dirty(&self) -> Option<AccountEntry> {
-		match *self {
-			AccountEntry::Cached(ref acc) if acc.is_dirty() => Some(AccountEntry::Cached(acc.clone_dirty())),
-			AccountEntry::Killed => Some(AccountEntry::Killed),
-			_ => None,
+		match self.is_dirty() {
+			true => Some( AccountEntry {
+				account: self.account.clone_dirty(),
+				state: self.state,
+			}),
+			false => None,
 		}
 	}
 
 	/// Clone account entry data that needs to be saved in the snapshot.
 	/// This includes basic account information and all locally cached storage keys
 	fn clone_for_snapshot(&self) -> AccountEntry {
-		match *self {
-			AccountEntry::Cached(ref acc) => AccountEntry::Cached(acc.clone_all()),
-			AccountEntry::Killed => AccountEntry::Killed,
-			AccountEntry::Missing => AccountEntry::Missing,
+		AccountEntry {
+			account: self.account.clone_for_snapshot(),
+			state: self.state,
+		}
+	}
+
+	fn new_dirty(account: AccountData) -> AccountEntry {
+		AccountEntry {
+			account: account,
+			state: AccountState::Dirty,
+		}
+	}
+	
+	fn new_clean(account: AccountData) -> AccountEntry {
+		AccountEntry {
+			account: account,
+			state: AccountState::Clean,
+		}
+	}
+
+	fn merge_snapshot(&mut self, other: AccountEntry) {
+		self.state = other.state;
+		match other.account {
+			AccountData::Missing => self.account = AccountData::Missing,
+			AccountData::Cached(acc) => match self.account {
+				AccountData::Cached(ref mut ours) => {
+					ours.revert_storage_changes();
+					ours.merge_with(acc);
```

Comment:
When reverting changes from a snapshot it makes sense to keep cached storage around and discard modified data.